lib/part.ml

```diff
@@ -0,0 +1,197 @@
+        | Some g ->
+            let sep_indent = sep_indent_of_groups g in
+            let next_part_name = part_name_of_groups g in
+            Part_begin (next_part_name, sep_indent)
+        | None ->
+          (match Re.exec_opt Regexp.part_end line with
+          | Some _ -> Part_end
+          | None -> Normal line))
+
+  let input_line_err i = match input_line i with
+    | exception End_of_file -> Error `End_of_file
+    | line -> Ok line
+
+  (* Once support for [@@@ parts] will be dropped `parse_part` should be much simpler *)
+  let rec parse_parts input make_part current_part part_lines nline =
+    let open Util.Result.Infix in
+    let nline = nline + 1 in
+    let line = input_line_err input in
+    match parse_line line, current_part with
+    | Normal line, _ ->
+      parse_parts input make_part current_part (line :: part_lines) nline
+    | Part_end, Some _ ->
+      parse_parts input (anonymous_part) None [] nline
+        >>| List.cons (make_part ~is_begin_end_part:true part_lines)
+    | Part_end, None -> Error ("There is no part to end.", nline)
+    | Part_begin (next_part_name, sep_indent), None ->
+        let next_part = next_part ~name:next_part_name ~sep_indent in
+        let rcall = parse_parts input next_part (Some next_part_name) [] nline in
+        if part_lines = [] then
+          rcall (* Ignore empty anonymous parts: needed for legacy support *)
+        else
+          rcall >>| List.cons (make_part ~is_begin_end_part:true part_lines)
+    | Compat_attr (name, sep_indent), None ->
+      let next_part = next_part ~name ~sep_indent in
+      parse_parts input next_part None [] nline
+        >>| List.cons (make_part ~is_begin_end_part:false part_lines)
+    | Part_begin _, Some p | Compat_attr _, Some p   ->
+      let msg = Printf.sprintf  "Part %s has no end." p in
+      Error (msg, nline)
+    | File_end, Some p ->
+        let msg = Printf.sprintf "File ended before part %s." p in
+        Error (msg, nline)
+    | File_end, None -> Ok [make_part ~is_begin_end_part:true part_lines]
+
+  let of_file name =
+    let input = open_in name in
+    match parse_parts input (anonymous_part) None [] 0 with
+    | Ok parts -> parts
+    | Error (msg, line) -> Fmt.failwith "In file %s, line %d: %s" name line msg
+end
+
+type file = Part.t list
+
+let read file = Parse_parts.of_file file
+
+let find file ~part = match part with
+  | Some part ->
+    (match List.find_opt (fun p -> String.equal (Part.name p) part) file with
+     | Some p -> Some [Part.body p]
+     | None   -> None )
+  | None      ->
+    List.fold_left (fun acc p -> Part.body p :: acc) [] file
+    |> List.rev
+    |> fun x -> Some x
+
+let rec replace_or_append part_name body = function
+  | p :: tl when String.equal (Part.name p) part_name ->
+    { p with body } :: tl
+  | p :: tl ->
+    p :: replace_or_append part_name body tl
+  | [] ->
+    [{ name = part_name; sep_indent = ""; body }]
+
+let replace file ~part ~lines =
```

Comment:
Is this still called somewhere? If we removed the `to-ml` direction this is probably deadcode. We can take care of it later but I'm writing it down so I will remember!

Reply:
Oh, and if the `sep_indent` thing is also only used by the `to-ml` direction it will be a very nice thing to cut^^